code_tools/2 type_checkers/index.md

```diff
@@ -0,0 +1,570 @@
+
+</details>
+
+1. Annotate the `factorial` function below:
+
+
+        def factorial(num):
+            total = 1
+            for i in range(2, num + 1):
+                total *= i
+            return total
+
+
+    <textarea name="form[q1]" rows="5" required=""></textarea>
+
+<details markdown="1"><summary  markdown="span">Generics</summary>
+
+Integers, floats, booleans and strings are primitive data types. Built into the language, they serve as building blocks for more complex data structures. For instance, you might need a `list` to store your data. 
+
+
+    numbers: list = [1, 2, 3]
+    number = numbers.pop()
+
+
+Here is the catch, the type `list` does not tell *anything* about what is in the `list`. So really what we have here is a `list` containing `Any`. In this case the type of `number` would be `Any` too.
+
+A `list` is a generic data type. It can store various types, but its operation will vary based on what you store. Simply put for a `list`, if you initially store integers in the list, you will later be able to retrieve integers from that list. This can be annotated as follows:
+
+
+    numbers: list[int] = [1, 2, 3]
+    number = numbers.pop()
+
+
+Now `numbers` is defined as a list of integers, and through that `number` will be of type `int` too.
+
+Let's take a quick look at `dict`. Dictionaries are generic over two types, their keys and values. This is how that can be annotated:
+
+
+    grades: dict[str, int] = {"Martijn": 7, "Marleen": 8}
+
+
+Tuples are an immutable data structure, once initialized it cannot be changed. So it is known up front exactly what the type of each value in the tuple is going to be. Because of this the `tuple` type can a variable amount of generic anotations with exactly as many types as there are values. Like so:
+
+    foo: tuple[int, float] = (7, 7.2)
+    bar: tuple[int, float, str] = (8, 7.9, "hello world")
+    baz: tuple[int, int, int] = (1, 2, 3)
+
+
+What about nested data structures?
+
+
+    stats: dict[str, tuple[int, float]] = {"Martijn": (7, 7.2), "Marleen": (8, 8.1)}
+
+
+Again, in most situations `mypy` can infer the types of the variables, and it is not strictly needed to annotate each data structure for type checking. That said, especially when it comes to data structures, annotations make the code easier to understand.  
+
+</details>
+
+2. Annotate the data structures below:
+
+        
+        foo = ["hello", "world"]
+    
+
+    <textarea name="form[q2.1]" rows="1" required=""></textarea>
+
+        
+        bar = [("Martijn", 1), ("Marleen", 2)]
+    
+
+    <textarea name="form[q2.2]" rows="1" required=""></textarea>
+
+        
+        baz = {1: {2: {3: "hello"}}}
+    
+
+    <textarea name="form[q2.3]" rows="1" required=""></textarea>
+
+<details markdown="1"><summary  markdown="span">Type variables</summary>
+
+Generics work through the use of type variables. In Python these variables are provided by `TypeVar` from the `typing` module. Here is how it works:
```

Comment:
"Generics work through the use of type variables."

Dit klinkt als een algemene uitleg, maar hieronder zie ik vooral hoe je kunt specificeren dat je een algemeen bruikbare functie schrijft terwijl je wel type hints wil gebruiken. En dat laatste lijkt me ook helemaal perfect, een mooie afsluiter.

Daarom zou ik het introduceren door meer het probleem te stellen: hoe schrijf je type hints voor een "generieke" functie?